main/java/mjc/Runner.java

```diff
@@ -0,0 +1,184 @@
+        System.out.println("Find FIRST weekday for product " + purchase2.getProduct() + ": " + findMapValue(purchasesFirstDayMap, purchase2));
+
+        removeMapEntries(purchasesLastDayMap, new EntryChecker<>() {
+            @Override
+            public boolean check(Map.Entry<AbstractPurchase, Weekday> entry) {
+                return entry.getKey().getProduct().getName().equals("meat");
+            }
+        });
+        removeMapEntries(purchasesFirstDayMap, new EntryChecker<>() {
+            @Override
+            public boolean check(Map.Entry<AbstractPurchase, Weekday> entry) {
+                return entry.getValue().equals(Weekday.FRIDAY);
+            }
+        });
+
+        System.out.println("\nRemoved entries where name is 'meat'");
+        printMap(purchasesLastDayMap);
+        System.out.println("Removed entries where weekday is 'FRIDAY'");
+        printMap(purchasesFirstDayMap);
+
+        System.out.println("Total cost of discounted purchases: " + discountedPurchases.getTotalCost() + "\n");
+
+        System.out.println("Weekday => List of purchases");
+        printMap(weekdayPurchaseEnumMap);
+
+        System.out.println("Total cost of purchases in a single day");
+        for (Map.Entry<Weekday, PurchaseList> entry : weekdayPurchaseEnumMap.entrySet()) {
+            System.out.println(entry.getKey() + " " + entry.getValue().getTotalCost());
+        }
+
+        System.out.println("\nFind purchases on MONDAY");
+        System.out.println(findMapValue(weekdayPurchaseEnumMap, Weekday.MONDAY));
+
+        removeMapEntries(weekdayPurchaseEnumMap, new EntryChecker<>() {
+            @Override
+            public boolean check(Map.Entry<Weekday, PurchaseList> entry) {
+                return entry.getValue().getPurchases().removeIf(purchase -> purchase.getProduct().getName().equals("milk"));
+            }
+        });
+        System.out.println("Removed entries where name is 'milk'");
+        printMap(weekdayPurchaseEnumMap);
+    }
+
+    public interface EntryChecker<K, V> {
+        boolean check(Map.Entry<K, V> entry);
+    }
+
+    private static <K, V> void removeMapEntries(Map<K, V> map, EntryChecker<K, V> checker) {
+        map.entrySet().removeIf(checker::check);
+    }
+
+    private static void putFirstDay(Map<AbstractPurchase, Weekday> map, Weekday day, AbstractPurchase purchase) {
+        if (!map.containsKey(purchase)) {
+            map.put(purchase, day);
+        }
+    }
+
+    private static void putLastDay(Map<AbstractPurchase, Weekday> map, Weekday day, AbstractPurchase purchase) {
+        map.remove(purchase);
+        map.put(purchase, day);
+    }
+
+    private static void putEnumMapPurchase(Map<Weekday, PurchaseList> map, Weekday day, AbstractPurchase purchase) {
+        if (map.containsKey(day)) {
+            map.get(day).add(purchase);
+        } else {
+            map.put(day, new PurchaseList(new ArrayList<>(List.of(purchase))));
+        }

```

Comment:
ok but there is better idiom. 
see [here](https://docs.oracle.com/javase/tutorial/collections/interfaces/map.html), section Multimaps at the end.